c_src/enif_jq.c

```diff
@@ -65,59 +83,82 @@ static ERL_NIF_TERM make_ok_return(ErlNifEnv* env, ERL_NIF_TERM result) {
-    ErlNifBinary erl_jq_filter, erl_json_text;
+    ErlNifBinary erl_jq_filter;
+    ErlNifBinary erl_json_text;
     if (!enif_inspect_binary(env, argv[0], &erl_jq_filter) ||
         !enif_inspect_binary(env, argv[1], &erl_json_text)) {
         ret = JQ_ERROR_BADARG;
-        strcpy(err_msg, "invalid input or filter");
+        const char* error_message = "Expected arguments of type binary but got something else";
+        error_msg_bin =
+            make_error_msg_bin(env, error_message, strlen(error_message));
         goto out;
-	}
+    }
 
     // ------------------------- parse input json -----------------------------
-    jv jv_json_text = jv_parse_sized((const char*)erl_json_text.data, erl_json_text.size);
+    // It is reasonable to assume that jv_parse_sized would not require the 
+    // input string to be NULL terminated. Unfortunately this is not the case.
+    // Error reporting depends on that the input string is NULL terminated
+    // so we have to copy the input to a larger memory block to make sure it
+    // is NULL terminated. A binary is used because small binaries are cheap
+    // as they can be allocated on the process heap.
+    ERL_NIF_TERM json_input;
+    memcpy(enif_make_new_binary(env, erl_json_text.size + 1, &json_input), erl_json_text.data, erl_json_text.size);
+    enif_inspect_binary(env, json_input, &erl_json_text);
+    erl_json_text.data[erl_json_text.size-1] = '\0';
+    jv jv_json_text = jv_parse_sized((const char*)erl_json_text.data, erl_json_text.size - 1);
     if (!jv_is_valid(jv_json_text)) {
         ret = JQ_ERROR_PARSE;
         // jv_invalid_get_msg destroys input jv object and returns new jv object
         jv_json_text = jv_invalid_get_msg(jv_json_text);
-        strncpy(err_msg, jv_string_value(jv_json_text),
-            MAX_ERR_MSG_LEN);
+        error_msg_bin =
+            make_error_msg_bin(env,
+                               jv_string_value(jv_json_text),
+                               jv_string_length_bytes(jv_copy(jv_json_text)));
         goto out;
     }
     //jv_dump(jv_json_text, dumpopts);
 
     // -------------------------- compile filter -------------------------------
-    if (erl_jq_filter.size >= MAX_JQ_FILTER_LEN) {
-        ret = JQ_ERROR_SYSTEM;
-        strcpy(err_msg, "jq filter is too long");
-        goto out;
-    }
-    strncpy(jv_filter, (const char*)erl_jq_filter.data, erl_jq_filter.size);
-    if (!jq_compile(jq, jv_filter)) {
+    // jq_compile requires that the input program (filter) is a NULL terminated string.
+    // Unfortunately, that forces us to create a new larger container that
+    // we can put the program inside. We use an Erlang binary for the container
+    // as small binaries can be efficiently allocated on the heap.
+    ERL_NIF_TERM compile_input;
+    memcpy(enif_make_new_binary(env, erl_jq_filter.size + 1, &compile_input), erl_jq_filter.data, erl_jq_filter.size);
+    enif_inspect_binary(env, compile_input, &erl_jq_filter);
```

Comment:
Q: perhaps it's needed to check the result of this, like in the case [above](https://github.com/emqx/jq/pull/13/files#diff-a61eff3c40a2b099fc0ff5cc888bbf38f94a9649d54738fcbb2a7f0f6128b0c5R109)?

Reply:
The check above [above](https://github.com/emqx/jq/pull/13/files#diff-a61eff3c40a2b099fc0ff5cc888bbf38f94a9649d54738fcbb2a7f0f6128b0c5R109) is for checking that the input parameters has the right type (i.e., binaries). This call to `enif_inspect_binary` is just to get hold of the `ErlNifBinary` struct for the compile_input as we already know that it is a binary.